### PR TITLE
feat: update `tryFinalize()` to check finality on each L2 block

### PR DESCRIPTION
**Pull request by lesterli**
## Summary

This PR updates `finalizer.go` to check finality on each L2 block, `tryFinalize()` doesn't call on each L2 block b/c `onPendingSafeUpdate` only triggered upon a new L1 block.

* add an `FinalizerL2Interface` interface which has the function `L2BlockRefByNumber(ctx context.Context, num uint64) (eth.L2BlockRef, error)`
* update the function `NewFinalizer` to add this interface as its input parameter
* update all the codes that call the function `NewFinalizer` and pass an instance of the `FinalizerL2Interface` interface

## Test Plan
```
make lint-go
make build-go
cd op-node && make test
```

connect w FP e2e tests locally and see it passed

```
    op_test_manager.go:354: [17:55:15.448] Registered Finality Provider cceb4691002c79802720f1adb9573d66e44c18edcad2b26e59b80750eb14e2e8 for op-stack-l2-901
    base_test_manager.go:174: successfully submitted a BTC delegation
    base_test_manager.go:207: delegations are pending
    base_test_manager.go:227: delegations are active
    op_test_manager.go:430: [17:55:36.294] The test manager is running with 1 finality-provider(s)
    utils.go:100: Public randomness for fp cceb4691002c79802720f1adb9573d66e44c18edcad2b26e59b80750eb14e2e8 is successfully committed at height 64
    op_e2e_test.go:172: [17:56:31.538] Stopped the FP instance
    op_e2e_test.go:176: [17:56:31.538] last processed height 56
    op_e2e_test.go:189: [17:56:43.065] Test case 1: OP chain block finalized stuck at height 56
    op_e2e_test.go:198: [17:57:12.090] Restarted the FP instance
    op_e2e_test.go:202: [17:57:22.093] Test case 2: OP chain finality is recover, the latest finalized block height 95
    op_test_manager.go:480: Stopping test manager
--- PASS: TestFinalityStuckAndRecover (193.73s)
PASS
ok  	github.com/babylonchain/finality-provider/itest/opstackl2	201.464s
```

### CODE REVIEW

File: op-node/rollup/finality/finalizer.go [hunk 1]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
+			// If there are any L2 blocks that are not finalized, then we can't finalize the later ones.
+			// This is because we need to finalize the L2 blocks in order.
+			blockCount := int(fd.L2Block.Number - finalizedL2.Number)
+			l2Blocks := make(map[uint64]eth.L2BlockRef)
+			queryBlocks := make([]*cwclient.L2Block, blockCount)
+			for i := 0; i < blockCount; i++ {
+				blockNumber := uint64(i) + finalizedL2.Number + uint64(1)
+				l2Block, err := fi.l2Fetcher.L2BlockRefByNumber(fi.ctx, blockNumber)
+				if err != nil {
+					fi.emitter.Emit(rollup.EngineTemporaryErrorEvent{Err: fmt.Errorf("failed to check if on finalizing L2 chain, could not fetch block %d: %w", blockNumber, err)})
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: "failed to check if on finalizing L2 chain" - what does this mean? I don't think we should copy the "failed to check if on finalizing L1 chain" and modify. the context is quite different and this modified sentence doesn't make much sense to me

i think we could just explain what exactly happened "could not fetch block..."

also thinking it more, it shouldn't emit a temporary event. if `fd.L2Block.Number` is 60, we shouldn't expect L2 block 59 missing. this is a critical err

File: op-node/rollup/finality/finalizer.go [hunk 2]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
-			queryParams := &sdk.L2Block{
-				BlockHeight:    fd.L2Block.Number,
-				BlockHash:      fd.L2Block.Hash.String(),
-				BlockTimestamp: fd.L2Block.Time,
+			// If there are any L2 blocks that are not finalized, then we can't finalize the later ones.
+			// This is because we need to finalize the L2 blocks in order.
+			blockCount := int(fd.L2Block.Number - finalizedL2.Number)
+			l2Blocks := make(map[uint64]eth.L2BlockRef)
+			queryBlocks := make([]*cwclient.L2Block, blockCount)
+			for i := 0; i < blockCount; i++ {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this function is getting too big. to make it more readable, we can extract these logic and the for loop for creating the queryBlocks to a helper function

File: op-node/rollup/finality/finalizer.go [hunk 3]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
+				}
+				l2Blocks[blockNumber] = l2Block
+
+				queryBlocks[i] = &cwclient.L2Block{
+					BlockHeight:    l2Block.Number,
+					BlockHash:      l2Block.Hash.String(),
+					BlockTimestamp: l2Block.Time,
+				}
+				fi.log.Debug(
+					"babylon gadget query params",
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: "added a block to babylon gadget's query params"

File: op-node/rollup/finality/finalizer.go [hunk 4]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
-				finalizedDerivedFrom = fd.L1Block
+			// set finalized block(s)
+			if lastFinalizedBlockNumber != nil {
+				for finalizedBlockNumber := *lastFinalizedBlockNumber; finalizedBlockNumber > finalizedL2.Number; finalizedBlockNumber-- {
+					finalizedL2Block := l2Blocks[*lastFinalizedBlockNumber]
+					fi.log.Debug("set finalized block", "l2_block", finalizedL2Block)
+					finalizedL2 = finalizedL2Block
+					finalizedDerivedFrom = fd.L1Block
+				}
 			}
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: could you help me understand what you are trying to do here?

let's say:
- `finalizedL2.Number` is block 55
- `queryBlocks` has blocks from 56 to 60
- `lastFinalizedBlockNumber` is 60

then in this for loop, we:
- first set `finalizedL2` to 60
- then `finalizedBlockNumber` becomes 60-1=59
- then we check `finalizedBlockNumber > finalizedL2.Number` again
- which equals to `59 > 60`, it will exit the loop

which looks wrong to me b/c this loop will always be called once and exit

File: op-node/rollup/finality/finalizer.go [hunk 5]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
-			babylonFinalized, err := fi.babylonFinalityClient.QueryIsBlockBabylonFinalized(queryParams)
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
+			// query the last finalized block in the block range on Babylon
+			lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+			fi.log.Debug("babylon gadget query result", "babylon_finalized_block_number", lastFinalizedBlockNumber)
 
 			// If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
 			// for any other error types, emit a critical error event.
-			if err != nil && !errors.Is(err, sdk.ErrNoFpHasVotingPower) {
+			if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: thinking deeper on it, this check is wrong. see https://github.com/babylonchain/babylon-finality-gadget/issues/59

this is non-trivial change. so likely will need another PR

File: op-node/rollup/finality/finalizer.go [hunk 6]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
-				"babylon gadget query params",
-				"block_height", queryParams.BlockHeight,
-				"block_hash", queryParams.BlockHash,
-				"block_timestamp", queryParams.BlockTimestamp,
-			)
-			babylonFinalized, err := fi.babylonFinalityClient.QueryIsBlockBabylonFinalized(queryParams)
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
+			// query the last finalized block in the block range on Babylon
+			lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+			fi.log.Debug("babylon gadget query result", "babylon_finalized_block_number", lastFinalizedBlockNumber)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: i added log and saw

```
finalizer.go:277:           ERROR[07-16|14:50:41.686] babylon gadget query result              role=verifier babylon_finalized_block_number=0xc00280c1f0
finalizer.go:290:           ERROR[07-16|14:50:41.686] set finalized block                      role=verifier l2_block=0503f4..d019cd:11
```
not sure why the block number is 0xc00280c1f0

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: ok seems we are printing the pointer address instead of the value

File: op-node/rollup/finality/finalizer.go [hunk 7]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
+			// set finalized block(s)
+			if lastFinalizedBlockNumber != nil {
+				for finalizedBlockNumber := *lastFinalizedBlockNumber; finalizedBlockNumber > finalizedL2.Number; finalizedBlockNumber-- {
+					finalizedL2Block := l2Blocks[*lastFinalizedBlockNumber]
+					fi.log.Debug("set finalized block", "l2_block", finalizedL2Block)
+					finalizedL2 = finalizedL2Block
+					finalizedDerivedFrom = fd.L1Block
+				}
 			}
 			// keep iterating, there may be later L2 blocks that can also be finalized
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go:259]: I just realized that this is problematic as well. we shouldn't keep iterating if some blocks in the range is not finalized

it also seems our tests didn't cover such cases. I think we should add more test coverages to make sure our code is robust

File: op-node/rollup/finality/finalizer.go [hunk 8]
```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
...
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
+			// query the last finalized block in the block range on Babylon
+			lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+			fi.log.Debug("babylon gadget query result", "babylon_finalized_block_number", lastFinalizedBlockNumber)
 
 			// If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
 			// for any other error types, emit a critical error event.
-			if err != nil && !errors.Is(err, sdk.ErrNoFpHasVotingPower) {
+			if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
 				fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf("failed to check if block %d is finalized on Babylon: %w", fd.L2Block.Number, err)})
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this also has to change to log the range instead of one block

**Comment by bap2pecs** [comment]
updated to fix all comments ^

**Comment by bap2pecs** [comment]
@lesterli let's make sure we run all the commands in test plan since we have disabled CI for now.

I just found that `cd op-node && make test` fails

File: op-node/rollup/finality/finalizer.go [hunk 9]
```diff
@@ -300,6 +287,98 @@ func (fi *Finalizer) tryFinalize() {
...
+			"block_hash", queryBlocks[i].BlockHash,
+			"block_timestamp", queryBlocks[i].BlockTimestamp,
+		)
+	}
+
+	lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+
+	// TODO: we shouldn't skip on no voting power.
+	// https://github.com/babylonchain/babylon-finality-gadget/issues/59
+	if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
```
Comment (SebastianElvis) [inline op-node/rollup/finality/finalizer.go:350]: This seems not consistent with the logic in Babylon. Babylon considers a block not finalisable if no FP has voting power at this height ([ref](https://github.com/babylonchain/babylon/blob/fe247dd38a83ccbf940dabed5dafaa69698b9eb0/x/finality/keeper/tallying.go#L62-L65)). This is because no BTC can be slashed at this height anyway. 

But maybe we need to have a different strategy for L2s. It's sequencer's responsibility to keep having BTC stake such that users will trutst him. Once sequencer does not have any BTC stake backed then users might not trust him anyway, and the L2 might need to stall. Not entirely sure about this part.

Maybe for now we can skip to limit the impact of BTC staking on L2's liveness?

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go:350]: @SebastianElvis yes you are right. I realized this problem yesterday and created https://github.com/babylonchain/babylon-finality-gadget/issues/59 which I am gonna work on today

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go:350]: but the fix will be in a separate PR. this PR is to fix non-consecutive quorom issue due to the fact that `tryFinalize` doesn't call on every single L2 block

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go:350]: > Maybe for now we can skip to limit the impact of BTC staking on L2's liveness?

I'd vote not to skip for security concern I described in https://github.com/babylonchain/babylon-finality-gadget/pull/38#issuecomment-2231526118

the original purpose of adding this bypass is to consider cases when the BTC staking is not activated. but I missed the fact that voting power can get to zero again if all FPs got slashed.

so the right fix is to query BTC staking activation height instead of relying on voting power

Comment (SebastianElvis) [reply op-node/rollup/finality/finalizer.go:350]: Good point on https://github.com/babylonchain/babylon-finality-gadget/pull/38. Maybe the best thing to do here is to stall then. Let's document this using the comments in the code ?

File: op-node/rollup/finality/finalizer.go [hunk 10]
```diff
@@ -300,6 +287,98 @@ func (fi *Finalizer) tryFinalize() {
...
+			"block_hash", queryBlocks[i].BlockHash,
+			"block_timestamp", queryBlocks[i].BlockTimestamp,
+		)
+	}
+
+	lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+
+	// TODO: we shouldn't skip on no voting power.
+	// https://github.com/babylonchain/babylon-finality-gadget/issues/59
+	if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) && lastFinalizedBlockNumber != nil {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: https://github.com/babylonchain/babylon-finality-gadget/pull/60#discussion_r1692063765

File: go.mod [hunk 11]
```diff
@@ -4,7 +4,7 @@ go 1.21
 
 require (
 	github.com/andybalholm/brotli v1.1.0
-	github.com/babylonchain/babylon-finality-gadget v0.0.0-20240708032647-4fdeba25732d
+	github.com/babylonchain/babylon-finality-gadget v0.1.3-alpha.0.20240722061210-54480a6216c7
```
Comment (bap2pecs) [inline go.mod]: do we need to update this since the finality gadget PR  https://github.com/babylonchain/babylon-finality-gadget/pull/60 is merged?

File: op-node/rollup/finality/finalizer_test.go [hunk 12]
```diff
@@ -91,6 +94,13 @@ func TestEngineQueue_Finalize(t *testing.T) {
 		BlockTime:     1,
 		SeqWindowSize: 2,
 	}
+	babylonCfg := &rollup.BabylonConfig{
+		ChainID:         "chain-test",
+		ContractAddress: "bbn1eyfccmjm6732k7wp4p6gdjwhxjwsvje44j0hfx8nkgrm8fs7vqfsa3n3gc",
+		BitcoinRpc:      "https://rpc.ankr.com/btc",
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer_test.go]: if we use a mock url, will it fail? for now, I don't know whether this test actually depend on the RPC or not

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer_test.go]: ok i tested it with "https://rpc.this-is-a-mock.com/btc" and still works

in general, we should use mock values in such cases to be explicit that there is no external dependencies

similarly, if we need a mock number e.g. a block number, sth like 12345 is better than 20178212 b/c people looking at it will wonder whether this magic number is actually needed

File: op-node/rollup/finality/finalizer.go [hunk 13]
```diff
@@ -245,11 +245,13 @@ func (fi *Finalizer) tryFinalize() {
 	// go through the latest inclusion data, and find the last L2 block that was derived from a finalized L1 block
 	for _, fd := range fi.finalityData {
 		if fd.L2Block.Number > finalizedL2.Number && fd.L1Block.Number <= fi.finalizedL1.Number {
-			shouldContinue, err := tryFinalizeOnConsecutiveBlockQuorom(fd, fi, &finalizedL2, &finalizedDerivedFrom)
+			lastFinalizedBlockNumber, err := tryFinalizeOnConsecutiveBlockQuorom(fd, fi, &finalizedL2, &finalizedDerivedFrom)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: let's add a commnet here that this function modifies the two pointer arguments: finalizedL2 and finalizedDerivedFrom

I am reading it and already got confused and realized after digging into the code.

File: op-node/rollup/finality/finalizer.go [hunk 14]
```diff
@@ -245,11 +245,13 @@ func (fi *Finalizer) tryFinalize() {
...
 			if err != nil {
 				return
 			}
-			if !shouldContinue {
+			// some blocks in the queried range is not BTC finalized, stop iterating to honor consecutive quorom
+			// all queried blocks are BTC finalized, continue iterating
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this docstring seems to be in the wrong place

File: op-node/rollup/finality/finalizer.go [hunk 15]
```diff
@@ -345,13 +347,22 @@ func tryFinalizeOnConsecutiveBlockQuorom(
...
 	// https://github.com/babylonchain/babylon-finality-gadget/issues/59
 	if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
+		if lastFinalizedBlockNumber != nil {
+			// set finalized block(s)
+			finalizedL2Block := l2Blocks[*lastFinalizedBlockNumber]
+			fi.log.Debug("set finalized block", "l2_block", finalizedL2Block)
+			*finalizedL2 = finalizedL2Block
+			*finalizedDerivedFrom = fd.L1Block
+
+			return lastFinalizedBlockNumber, nil
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this is copied from below. we should avoid copying code as much as possible, sometimes even when there is one line of duplication.

this should be extracted to a helper function and call it here
```
return someFunc(...)
```

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: wait, this logic seems wrong. seems we are suppressing the error here again?

File: op-node/rollup/finality/finalizer_test.go (hunk 12 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer_test.go]: changed

File: op-node/rollup/finality/finalizer.go (hunk 13 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer.go]: refactor it

File: op-node/rollup/finality/finalizer.go (hunk 15 above)
Comment (lesterli) [reply op-node/rollup/finality/finalizer.go]: updated

File: op-node/rollup/finality/finalizer.go [hunk 16]
```diff
@@ -236,33 +245,22 @@ func (fi *Finalizer) tryFinalize() {
...
-			)
-			babylonFinalized, err := fi.babylonFinalityClient.QueryIsBlockBabylonFinalized(queryParams)
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
-
-			// If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
-			// for any other error types, emit a critical error event.
-			if err != nil && !errors.Is(err, sdk.ErrNoFpHasVotingPower) {
-				fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf("failed to check if block %d is finalized on Babylon: %w", fd.L2Block.Number, err)})
+			lastFinalizedBlock, err := fi.findFinalizedL2Block(fd.L2Block.Number, finalizedL2.Number)
+			if err != nil {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this confused me a bit b/c when there is err but we don't log it here. only after digging into the code, I see that it's already emitted inside `findFinalizedL2Block`

so given that we only need to check "where there is error", a better API design is to return a bool `hasErr`. wdyt?

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: also `findFinalizedL2Block` should be renamed to sth like `findLastFinalizedL2BlockWithConsecutiveQuorom`

File: op-node/rollup/finality/finalizer.go [hunk 17]
```diff
@@ -236,33 +245,22 @@ func (fi *Finalizer) tryFinalize() {
...
+				if lastFinalizedBlock != (eth.L2BlockRef{}) {
+					// set finalized block(s)
+					finalizedL2, finalizedDerivedFrom = fi.updateFinalized(lastFinalizedBlock, fd.L1Block)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: should we use `lastFinalizedBlock.L1Origin` here instead of `fd.L1Block`?

b/c `lastFinalizedBlock` might not be the same as `fd.L2Block.Number`

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: this is probably another edge case that test coverage failed. we should have a test case for this where fd.L1Block is different from lastFinalizedBlock.L1Origin

File: op-node/rollup/finality/finalizer.go [hunk 18]
```diff
@@ -236,33 +245,22 @@ func (fi *Finalizer) tryFinalize() {
...
-			// set finalized status
-			if babylonFinalized {
-				fi.log.Debug("set finalized status", "l2_block", fd.L2Block)
-				finalizedL2 = fd.L2Block
-				finalizedDerivedFrom = fd.L1Block
+			if lastFinalizedBlock == (eth.L2BlockRef{}) {
+				break
+			}
+			// set finalized block(s)
+			finalizedL2, finalizedDerivedFrom = fi.updateFinalized(lastFinalizedBlock, fd.L1Block)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: this is the same call as in L252. i think the overall logic from line 249-264 is not very clean

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: same for the `if lastFinalizedBlock == (eth.L2BlockRef{}) {` check. we checked it twice

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: we can do sth like this:

image

so we only need to call `updateFinalized` once; and check `lastFinalizedBlock != (eth.L2BlockRef{})` once

File: op-node/rollup/finality/finalizer.go (hunk 16 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/38f662595b76067e505493fd412697e2b2ae4b94

File: op-node/rollup/finality/finalizer.go (hunk 17 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/38f662595b76067e505493fd412697e2b2ae4b94

File: op-node/rollup/finality/finalizer.go (hunk 18 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/38f662595b76067e505493fd412697e2b2ae4b94

File: op-node/rollup/finality/finalizer.go [hunk 19]
```diff
@@ -236,33 +245,20 @@ func (fi *Finalizer) tryFinalize() {
...
-			babylonFinalized, err := fi.babylonFinalityClient.QueryIsBlockBabylonFinalized(queryParams)
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
-
-			// If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
-			// for any other error types, emit a critical error event.
-			if err != nil && !errors.Is(err, sdk.ErrNoFpHasVotingPower) {
-				fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf("failed to check if block %d is finalized on Babylon: %w", fd.L2Block.Number, err)})
+
+			if hasErr {
 				return
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: thinking deeper on it. "return on err" is inappropriate now. if you see the below logic where `if finalizedDerivedFrom != (eth.BlockID{}) {` is checked, whenever `finalizedDerivedFrom` is set, we have to do some sanity checks.

so we need to:
- remove the "return on error" logic
- `findLastFinalizedL2BlockWithConsecutiveQuorom` doesn't need to return `hasErr` now

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/b28870d27649531ca240608a03f48dc8084b0daa

File: op-node/rollup/finality/finalizer.go [hunk 20]
```diff
@@ -300,6 +292,80 @@ func (fi *Finalizer) tryFinalize() {
 	}
 }
 
+/*
+ *  findLastFinalizedL2BlockWithConsecutiveQuorom tries to find the finalized L2 block.
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: tries to find the last finalized L2 block with consecutive quorom

File: op-node/rollup/finality/finalizer.go [hunk 21]
```diff
@@ -300,6 +292,80 @@ func (fi *Finalizer) tryFinalize() {
...
+ *  If there are any L2 blocks that are not finalized, then we can't finalize the later ones.
+ *  This is because we need to finalize the L2 blocks in order to guarantee consecutive quorom.
+ *
+ *  It queries the Babylon gadget to check if the L2 blocks are finalized.
+ *  If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
+ *  for any other error types, emit an critical error event.
+ */
+func (fi *Finalizer) findLastFinalizedL2BlockWithConsecutiveQuorom(
+	fdL2BlockNumber uint64,
+	finalizedL2Number uint64,
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: let's add some comments to explain the return value and those arguments

File: op-node/rollup/finality/finalizer.go [hunk 22]
```diff
@@ -300,6 +292,80 @@ func (fi *Finalizer) tryFinalize() {
...
+		fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf(
+			"failed to check if block %d to %d is finalized on Babylon: %w",
+			finalizedL2Number+1,
+			fdL2BlockNumber,
+			err,
+		)})
+		return eth.L2BlockRef{}
+	}
+
+	return l2Blocks[*lastFinalizedBlockNumber]
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: what if `lastFinalizedBlockNumber` is nil?

File: op-node/rollup/finality/finalizer.go [hunk 23]
```diff
@@ -300,6 +292,80 @@ func (fi *Finalizer) tryFinalize() {
...
+ *  This is because we need to finalize the L2 blocks in order to guarantee consecutive quorom.
+ *
+ *  It queries the Babylon gadget to check if the L2 blocks are finalized.
+ *  If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
+ *  for any other error types, emit an critical error event.
+ */
+func (fi *Finalizer) findLastFinalizedL2BlockWithConsecutiveQuorom(
+	fdL2BlockNumber uint64,
+	finalizedL2Number uint64,
+) eth.L2BlockRef {
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: why not return a pointer?

i think using a pointer type will be cleaner so the check above 
```
if lastFinalizedBlock.Number != fd.L2Block.Number {
```

will be changed to 
```
if lastFinalizedBlock == nil || lastFinalizedBlock.Number != fd.L2Block.Number {
```

so it's more explicit

File: op-node/rollup/finality/finalizer.go [hunk 24]
```diff
@@ -300,6 +292,80 @@ func (fi *Finalizer) tryFinalize() {
...
+		)
+	}
+
+	lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+
+	// TODO: we shouldn't skip on no voting power.
+	// https://github.com/babylonchain/babylon-finality-gadget/issues/59
+	if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
+		if lastFinalizedBlockNumber != nil {
+			fi.log.Warn("Received error but also had the finalized block", "error", err, "blockNumber", *lastFinalizedBlockNumber)
```
Comment (bap2pecs) [inline op-node/rollup/finality/finalizer.go]: the logic here is not clean for the same reason that we have too many unnecessary nesting and condition checks

it can be simplified to be:

- if there is error, emit
- if last finalized block is not nil, return the block
- else return nil

Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/c89dd55766047361ef5e3c04d497be5a17aeee0d

File: op-node/rollup/finality/finalizer.go (hunk 23 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/c89dd55766047361ef5e3c04d497be5a17aeee0d

File: op-node/rollup/finality/finalizer.go (hunk 22 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/c89dd55766047361ef5e3c04d497be5a17aeee0d

File: op-node/rollup/finality/finalizer.go (hunk 21 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/c89dd55766047361ef5e3c04d497be5a17aeee0d

File: op-node/rollup/finality/finalizer.go (hunk 20 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go]: fixed in https://github.com/babylonchain/optimism/pull/23/commits/c89dd55766047361ef5e3c04d497be5a17aeee0d

**Review by bap2pecs — COMMENTED** [review]
given that this PR is already way too big. let's move all the changes in finalizer_test.go to another PR so we can start new discussion threads there

also the tests seem still failing

**Review by bap2pecs — APPROVED** [review]
ok I fixed all the comments above. @lesterli you can create a new PR to fix test. it's currently failing. I copied the code of `finalizer_test.go` to https://gist.github.com/bap2pecs/65e97276d3253e3973f138ab84878e74

I am gonna merge this PR now. I tested against FP e2e tests and all passed

let's keep the branch for now since I don't know if FP or SDK repo depends on some commits in this branch

File: op-node/rollup/finality/finalizer.go (hunk 9 above)
Comment (bap2pecs) [reply op-node/rollup/finality/finalizer.go:350]: after https://github.com/babylonchain/babylon-finality-gadget/issues/59, the chain will stall on any error querying the SDK. not just `ErrNoFpHasVotingPower`. added the comments in https://github.com/babylonchain/optimism/commit/20ce4dce5061ca0a0da70f6868c66bb939f9570d
